composer/trainer/dist_strategy.py

```diff
@@ -0,0 +1,262 @@
+            # Otherwise wrap if root object `obj.fsdp_wrap_fn(module)` is true
+            # Or if unwrapped params in module in greater than or equal to fsdp_config.min_params
+            def _auto_wrap_policy(module: torch.nn.Module, recurse: bool, unwrapped_params: int) -> bool:
+                min_params = int(fsdp_config.get('min_params', 1e8))
+                if recurse:
+                    return True
+                else:
+                    if hasattr(module, '_fsdp_wrap'):
+                        return bool(module._fsdp_wrap)
+
+                    is_large = unwrapped_params >= min_params
+                    if hasattr(obj, 'fsdp_wrap_fn') and isinstance(obj.fsdp_wrap_fn, Callable):
+                        return obj.fsdp_wrap_fn(module) or is_large
+                    else:
+                        return is_large
+
+            fsdp_obj = FullyShardedDataParallel(
+                obj,
+                sharding_strategy=sharding_strategy,
+                auto_wrap_policy=_auto_wrap_policy,
+                cpu_offload=cpu_offload,
+                mixed_precision=mixed_precision,
+                backward_prefetch=backward_prefetch,
+                param_init_fn=_param_init_fn,
+            )
+
+            # Activation Checkpointing
+            if fsdp_config.get('activation_checkpointing', False):
+                if fsdp_config.get('activation_checkpointing_offload_to_cpu', False):
+                    checkpoint_wrapper_fn = lambda module: checkpoint_wrapper(checkpoint_wrapper(module),
+                                                                              offload_to_cpu=True)
+                else:
+                    checkpoint_wrapper_fn = checkpoint_wrapper
+
+                # Choose which modules to activation checkpoint according to the following priority:
+                # If module has attribute `module._activation_checkpointing = ...`, always respect it
+                # Otherwise checkpoint if root object `obj.activation_checkpointing_fn(module)` is true
+                def _check_fn(module: torch.nn.Module) -> bool:
+                    if hasattr(module, '_activation_checkpointing'):
+                        return bool(module._activation_checkpointing)
+                    if hasattr(obj, 'activation_checkpointing_fn') and isinstance(obj.activation_checkpointing_fn,
+                                                                                  Callable):
+                        return obj.activation_checkpointing_fn(module)
+                    return False
+
+                apply_activation_checkpointing_wrapper(
+                    fsdp_obj,
+                    checkpoint_wrapper_fn=checkpoint_wrapper_fn,  # type: ignore
+                    check_fn=_check_fn,  # type: ignore
+                )
+
+            setattr(model, name, fsdp_obj)
+
+    # Print model arch for debugging
+    if fsdp_config.get('verbose', False):
+        print(model)
```

Comment:
Should this (debug mode print) stay here?

Reply:
Yeah I think this is pretty useful for users, would like to flesh it out a bit more but keep it here.